e2e/test/iothub/service/FileUploadNotificationE2ETest.cs

```diff
@@ -18,41 +16,103 @@
+
+            // Client should still be able to receive file upload notifications after being closed and re-opened.
+            await UploadFile().ConfigureAwait(false);
+            WaitForFileUploadNotification(ref fileUploadNotificationReceivedCount, 1);
+            await serviceClient.FileUploadNotificationProcessor.CloseAsync().ConfigureAwait(false);
+        }
+
+        [LoggedTestMethod, Timeout(TestTimeoutMilliseconds)]
+        public async Task FileUploadNotification_ReceiveMultipleNotificationsInOneConnection()
+        {
+            using var serviceClient = new IotHubServiceClient(TestConfiguration.IoTHub.ConnectionString);
+
+            int fileUploadNotificationReceivedCount = 0;
+            Func<FileUploadNotification, AcknowledgementType> OnFileUploadNotificationReceived = (fileUploadNotification) =>
+            {
+                fileUploadNotificationReceivedCount++;
+                return AcknowledgementType.Complete;
+            };
+
+            serviceClient.FileUploadNotificationProcessor.FileUploadNotificationProcessor = OnFileUploadNotificationReceived;
 
+            await serviceClient.FileUploadNotificationProcessor.OpenAsync().ConfigureAwait(false);
+            await UploadFile().ConfigureAwait(false);
+            await UploadFile().ConfigureAwait(false);
+
+            // The open file upload notification processor should be able to receive more than one
+            // file upload notification without closing and re-opening as long as there is more
+            // than one file upload notification to consume.
+            WaitForFileUploadNotification(ref fileUploadNotificationReceivedCount, 2);
             await serviceClient.FileUploadNotificationProcessor.CloseAsync().ConfigureAwait(false);
-            fileUploaded.Should().BeTrue();
         }
 
-        private AcknowledgementType fileUploadCallback(FileUploadNotification notification)
+        /// <summary>
+        /// Wait until the expected number of file upload notifications have been received. If the expected
+        /// number of notifications are not received in time, this method throws a AssertionFailedException.
+        /// </summary>
+        /// <param name="fileUploadNotificationReceivedCount">The current number of file upload notifications received.</param>
+        /// <param name="expectedFileUploadNotificationReceivedCount">The expected number of file upload notifications to receive in this test.</param>
+        private void WaitForFileUploadNotification(ref int fileUploadNotificationReceivedCount, int expectedFileUploadNotificationReceivedCount)
         {
-            fileUploaded = true;
-            return AcknowledgementType.Complete;
+            var timer = Stopwatch.StartNew();
+            while (fileUploadNotificationReceivedCount < expectedFileUploadNotificationReceivedCount && timer.ElapsedMilliseconds < 60000)
+            {
+                continue;
```

Comment:
No need for a sleep here? Seems this could throttle the CPU real fast.

Reply:
Good idea